pkg/cloud/services/gc/ec2.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	rgapi "github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+)
+
+func (s *Service) deleteEC2Resources(ctx context.Context, resources []*rgapi.ResourceTagMapping) error {
+	for i := range resources {
+		res := resources[i]
+
+		parsedARN, err := arn.Parse(*res.ResourceARN)
+		if err != nil {
+			return fmt.Errorf("parsing arn %s: %w", *res.ResourceARN, err)
+		}
+
+		if strings.HasPrefix(parsedARN.Resource, "security-group/") {
+			s.scope.V(2).Info("Deleting Security group", "arn", parsedARN.String())
+			return s.deleteSecurityGroup(ctx, &parsedARN, res)
+		}
+	}
+
+	s.scope.V(2).Info("Finished deleting ec2 resources")
+
+	return nil
+}
+
+func (s *Service) deleteSecurityGroup(ctx context.Context, lbARN *arn.ARN, mapping *rgapi.ResourceTagMapping) error {
+	eksClusterName := getTagValue(eksClusterNameTag, mapping)
+	if eksClusterName != "" {
+		s.scope.V(2).Info("Security group created by EKS directly, skipping deletion", "cluster_name", eksClusterName)
+
+		return nil
+	}
+
+	//TODO: should we check for the security group name start with k8s-elb-
```

Comment:
I find it confusing that the caller filters the set of resources to pass to the clean up function, e.g., https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3610/files#diff-ac45a93b8f5e87ab26c8de0dd3165846262b1cb89b0e318586f6ed2b33f61627R127, and then the clean up function filters the set further.

Can we make either the caller, or this function, responsible for all the filtering?

Reply:
The caller is only filtering based on the resource type (i.e. `security-group`). I'm not sure this will be any clearer:

```go
if strings.HasPrefix(parsedARN.Resource, "security-group/") {
  eksClusterName := getTagValue(eksClusterNameTag, mapping)
  if eksClusterName != "" {
    s.scope.V(2).Info("Security group created by EKS directly, skipping deletion", "cluster_name", eksClusterName)
     continue
  }
  s.scope.V(2).Info("Deleting Security group", "arn", parsedARN.String())
  return s.deleteSecurityGroup(ctx, &parsedARN, res)
}
```

But happy to change if you want.

Reply:
I think what I am trying to say is this: if I had to call this function, I'm not how I would know that my input is correct.

Right now, the caller needs to filter resources before calling this function, but I think that's not enforced by the compiler.

For example, what happens if I call the function as it currently stands and pass a resource that is NOT a security group? Does `groupID` get an invalid value, and then the `DeleteSecurityGroupWithContext` call fails?

Reply:
Ah, i see what you are saying. I guess i am relying on the fact that the only exportable function is `ReconcileDelete` and from here we control what calls what. 

The current call to `deleteSecurityGroup` is protected with a check (i.e. `strings.HasPrefix(parsedARN.Resource, "security-group/") `). But what you are saying is that someone could come along later and make a change in the `gc` package somewhere and call `deleteSecurityGroup` without this check.

Its a bit late in the day for me...will ponder this tomorrow.

Reply:
@dlipovetsky - i have changed this now so that the clean up functions get passed the list of all the resources, they then decide to filter the resources and delete if required.